datafusion/ffi/src/record_batch_stream.rs

```diff
@@ -0,0 +1,190 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::{ffi::c_void, sync::Arc, task::Poll};
+
+use abi_stable::{
+    std_types::{ROption, RResult, RString},
+    StableAbi,
+};
+use arrow::array::{Array, RecordBatch};
+use arrow::{
+    array::{make_array, StructArray},
+    datatypes::{Schema, SchemaRef},
+    ffi::{from_ffi, to_ffi, FFI_ArrowArray, FFI_ArrowSchema},
+};
+use async_ffi::{ContextExt, FfiContext, FfiPoll};
+use datafusion::error::Result;
+use datafusion::{
+    error::DataFusionError,
+    execution::{RecordBatchStream, SendableRecordBatchStream},
+};
+use futures::{Stream, TryStreamExt};
+
+#[repr(C)]
+#[derive(Debug, StableAbi)]
+pub struct WrappedArray {
+    #[sabi(unsafe_opaque_field)]
+    array: FFI_ArrowArray,
+
+    schema: WrappedSchema,
+}
+
+#[repr(C)]
+#[derive(Debug, StableAbi)]
+pub struct WrappedSchema(#[sabi(unsafe_opaque_field)] FFI_ArrowSchema);
+
+impl From<SchemaRef> for WrappedSchema {
+    fn from(value: SchemaRef) -> Self {
+        let schema = FFI_ArrowSchema::try_from(value.as_ref());
+        WrappedSchema(schema.unwrap_or(FFI_ArrowSchema::empty()))
```

Comment:
I don't really like that this silently fails but it seems like the only error it could realistically encounter is https://github.com/apache/arrow-rs/blob/1666a4d7b237370c167e2b90b45d1354e28f500a/arrow-schema/src/ffi.rs#L745-L747.